File: scripts/reflecta.py
import os, sys, json, hashlib, builtins

if sys.implementation.name == "micropython":
    import array, binascii, builtins, cmath, collections, errno, heapq, io, math, re, select, struct, btree, framebuf, uctypes, micropython
elif sys.implementation.name == "cpython":
    import importlib, ctypes
    for n in sys.builtin_module_names:
        globals()[n] = importlib.import_module(n)

class Reflecta:
    fuzzout = None
    modtype = type(builtins)

    def crash():
        if sys.implementation.name == "cpython":
            os.kill(os.getpid(), 9)
        elif sys.implementation.name == "micropython":
            import machine
            machine.Signal(9).on()

    def print(s):
        if Reflecta.fuzzout is None:
            if os.getenv("fuzzout"):
                Reflecta.fuzzout = open(os.getenv("fuzzout"), "w")
            else:
                Reflecta.fuzzout = sys.stdout

        Reflecta.fuzzout.write(str(s) + "\n")
        Reflecta.fuzzout.flush()

    def enumerate():
        paths = []
        for m in globals().values():
            if type(m) == Reflecta.modtype:
                paths.extend(Reflecta.builtins(m, m.__name__))

        for p in paths:
            b = eval(p)
            print(p, Reflecta.kindof(b))

        Reflecta.print(json.dumps(paths))


    def reflect(obj, index):
        return {
            "index": index,
            "kind": Reflecta.kindof(obj),
            "type": Reflecta.typeof(obj),
            "arity": Reflecta.arityof(obj),
            "fields": Reflecta.fieldsof(obj),
            "methods": Reflecta.methodsof(obj),
        }

    def record(obj, index):
        Reflecta.print(json.dumps(Reflecta.reflect(obj, index)))

    def filter(b):
        try:
            return Reflecta.kindof(eval(b)) != "primitive"
        except:
            return False

    def builtins(m, path, visited=set()):
        visited.add(id(m))

        paths = [path]
        for n in dir(m):
            attr = getattr(m, n)
            if id(attr) in visited or n.startswith("_"):
                continue
            elif isinstance(attr, Reflecta.modtype):
                paths.extend(Reflecta.builtins(attr, path + "." + n))
            else:
                paths.append(path + "." + n)

        return [p for p in paths if Reflecta.filter(p)]

    def kindof(obj):
        if type(obj) in [builtins.str, builtins.int, builtins.float]:
            return "primitive"
        elif type(obj).__name__ == "type":
            return "class"
        elif "function" in type(obj).__name__ or callable(obj):
            return "function"
        else:
            return "object"

    def typeof(obj):
        kind = Reflecta.kindof(obj)
        if kind == "primitive":
            return type(obj).__name__
        elif kind in ["class", "function"]:
            if hasattr(obj, "__name__") and obj.__name__:
                return obj.__name__
            else:
                return kind
        else:
            if hasattr(obj, "__class__") and hasattr(obj.__class__, "__name__"):
                return obj.__class__.__name__

    def fieldsof(obj):
        return dir(obj)

    def methodsof(obj):
        return [(m, "null") for m in dir(obj) if Reflecta.kindof(getattr(obj, m)) == "function"]

    def arityof(obj):
        return None

# Reflecta.enumerate()
# try:
#     raise Exception()
# except:
#     pass

File: scripts/reflecta.js
class Reflecta {
    static crash() {
        fuzzilli('FUZZILLI_CRASH', 0)
    }

    static builtins(obj = globalThis, path = "globalThis", visited = new Set()) {
        visited.add(obj)

        let paths = [path]
        for (let p of Object.getOwnPropertyNames(obj)) {
            if (visited.has(obj[p]) || !p.match(/^\w+$/) || ["prototype", "length", "arguments", "caller", "name", "Reflecta"].includes(p)) {
                continue
            } else if (Reflecta.kindof(obj[p]) === "primitive") {
                paths.push(path + "." + p)
            } else {
                paths.push(...Reflecta.builtins(obj[p], path + "." + p, visited))
            }
        }

        return paths
    }

    static enumerate() {
        let builtins = Reflecta.builtins().map((b) => b.replace("globalThis.", ""))
        let json = JSON.stringify(builtins)
        Reflecta.print(json)
    }

    static print(str) {
        fuzzilli('FUZZILLI_PRINT', str)
    }

    static record(obj, index) {
        Reflecta.print(JSON.stringify(Reflecta.reflect(obj, index)))
    }

    static reflect(obj, index) {
        return {
            "index": index,
            "kind": Reflecta.kindof(obj),
            "type": Reflecta.typeof(obj),
            "arity": Reflecta.arityof(obj),
            "fields": Reflecta.fieldsof(obj),
            "methods": Reflecta.methodsof(obj),
        }
    }

    static kindof(obj) {
        switch (typeof obj) {
            case "string":
            case "number":
            case "boolean":
            case "bigint":
            case "symbol":
            case "null":
            case "undefined":
                return "primitive"
            case "object":
                return obj === null ? "primitive" : "object"
            case "function":
                if (Object.hasOwn(obj, "prototype") && Object.hasOwn(obj, "name") && obj.name.match(/^[A-Z]\w*$/))
                    return "class"
                else
                    return "function"
            default:
                throw 0
        }
    }

    static typeof(obj) {
        switch (Reflecta.kindof(obj)) {
            case "primitive":
                return (typeof obj)
            case "object":
                if (Object.hasOwn(obj, "constructor") && Object.hasOwn(obj.constructor, "name")) {
                    return "object." + obj.constructor.name
                } else {
                    return "object"
                }
            case "class":
                if (Object.hasOwn(obj, "name")) {
                    return "class." + obj.name
                } else {
                    return "class"
                }
            case "function":
                if (Object.hasOwn(obj, "name")) {
                    return "function." + obj.name
                } else {
                    return "function." + Reflecta.arityof(obj)
                }
        }
    }

    static methodsof(obj) {
        return Object
            .getOwnPropertyNames(obj ?? {})
            .filter((p) => Reflecta.kindof(obj[p]) === 'function')
            .map((p) => [p, Reflecta.arityof(obj[p]).toString()])
    }

    static arityof(fn) {
        if (["class", "function"].includes(Reflecta.kindof(fn)) && Object.hasOwn(fn, "length")) {
            return fn.length >= 0 ? fn.length : null
        } else {
            return null
        }
    }

    static fieldsof(obj) {
        return Object.getOwnPropertyNames(obj ?? {})
    }
}
